File: core/typed_ir/typed_topology_ir.py
from __future__ import annotations

from typing import Any, Dict, List

from core.runtime.distributed_topology_engine import build_distributed_topology

from .schema_types import SemanticNode, SemanticEdge


def compile_typed_topology_ir(services: List[str]) -> Dict[str, Any]:
    topo = build_distributed_topology(sorted(services))
    nodes: List[SemanticNode] = []
    edges: List[SemanticEdge] = []
    for n in topo.get("nodes", []):
        nodes.append(SemanticNode(id=str(n.get("id", "")), type=str(n.get("type", "service"))))
    for e in topo.get("edges", []):
        edges.append(
            SemanticEdge(
                source=str(e.get("from", "")),
                target=str(e.get("to", "")),
                relation=str(e.get("relation", "distributed_dependency")),
                evidence=["distributed_topology"],
            )
        )
    return {
        "nodes": nodes,
        "edges": edges,
        "topology": topo,
        "typed": True,
        "deterministic": True,
    }

File: src/typed_ir/typedTopologyIr.ts
/**
 * Converted from Python: core/typed_ir/typed_topology_ir.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";
import { buildDistributedTopology } from "../runtime/distributedTopologyEngine.js";
import { SemanticNode, SemanticEdge } from "./schemaTypes.js";

export function compileTypedTopologyIr(services: any): any {
  var topo: any = buildDistributedTopology(py.sorted(services));
  var nodes: any[] = [];
  var edges: any[] = [];
  var n: any;
  for (n of py.iter(py.get(topo, "nodes", []))) {
    py.listAppend(nodes, new SemanticNode(py.toStr(py.get(n, "id", "")), py.toStr(py.get(n, "type", "service"))));
  }
  var e: any;
  for (e of py.iter(py.get(topo, "edges", []))) {
    py.listAppend(edges, new SemanticEdge(py.toStr(py.get(e, "from", "")), py.toStr(py.get(e, "to", "")), py.toStr(py.get(e, "relation", "distributed_dependency")), ["distributed_topology"]));
  }
  return {"nodes": nodes, "edges": edges, "topology": topo, "typed": true, "deterministic": true};
}
export { SemanticEdge, SemanticNode, buildDistributedTopology };
